deepsubtree_test.go

```diff
@@ -232,18 +240,165 @@ func TestDeepSubtreeWWithAddsAndDeletes(t *testing.T) {
 
 	// Delete all the keys we added and check root hashes stay equal
 	for i := range keysToAdd {
-		keyToAdd := keysToAdd[i]
-		dst.Remove(keyToAdd)
-		dst.SaveVersion()
-		rootHash, err := dst.WorkingHash()
+		keyToDelete := keysToAdd[i]
+
+		existenceProofs, err := tree.getExistenceProofsNeededForRemove(keyToDelete)
+		require.NoError(err)
+		rootHash, err := tree.WorkingHash()
 		require.NoError(err)
-		err = dst.BuildTree(rootHash)
+		err = dst.AddExistenceProofs(existenceProofs, rootHash)
 		require.NoError(err)
-		tree.Remove(keyToAdd)
+
+		dst.Remove(keyToDelete)
+		dst.SaveVersion()
+		tree.Remove(keyToDelete)
 		tree.SaveVersion()
 
 		areEqual, err := haveEqualRoots(dst.MutableTree, tree)
 		require.NoError(err)
 		require.True(areEqual)
 	}
 }
+
+func readByte(r *bytes.Reader) byte {
+	b, err := r.ReadByte()
+	if err != nil {
+		return 0
+	}
+	return b
+}
+
+func FuzzBatchAddReverse(f *testing.F) {
+	f.Fuzz(func(t *testing.T, input []byte) {
+		require := require.New(t)
+		if len(input) < 1000 {
+			return
+		}
+		tree, err := NewMutableTreeWithOpts(db.NewMemDB(), cacheSize, nil, true)
+		require.NoError(err)
+		dst := NewDeepSubTree(db.NewMemDB(), cacheSize, true, 0)
+		r := bytes.NewReader(input)
+		keys := make(set.Set[string])
+		// Generates random new key half times and an existing key for the other half times.
+		key := func(tree *ImmutableTree, genRandom bool) (isRandom bool, key []byte) {
+			if genRandom && readByte(r) < math.MaxUint8/2 {
+				k := make([]byte, readByte(r)/2)
+				r.Read(k)
+				val, err := tree.Get(k)
+				require.NoError(err)
+				if len(k) == 0 || val != nil {
```

Comment:
Do we need to care about empty keys? What about using `randByte(r)/2+1` as `k` size? Code is `continue`-ing anyways if returned key is `nil`

Reply:
Using `randbyte(r)/2 + 1` makes sense. Thanks for the suggestion.